indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/symmetric/SymmetricKey.kt

```diff
@@ -0,0 +1,211 @@
+
+        override fun equals(other: Any?): Boolean {
+            if (this === other) return true
+            if (other !is WithDedicatedMac<*>) return false
+
+            if (algorithm != other.algorithm) return false
+            if (!encryptionKey.contentEquals(other.encryptionKey)) return false
+            if (!macKey.contentEquals(other.macKey)) return false
+            if (additionalProperties != other.additionalProperties) return false
+
+            return true
+        }
+
+        override fun hashCode(): Int {
+            var result = algorithm.hashCode()
+            result = 31 * result + encryptionKey.contentHashCode()
+            result = 31 * result + macKey.contentHashCode()
+            return result
+        }
+    }
+}
+
+/**Use to smart cast*/
+@OptIn(ExperimentalContracts::class)
+fun <A : AuthCapability<K>, K : KeyType, I : NonceTrait> SymmetricKey<A, I, K>.isAuthenticated(): Boolean {
+    contract {
+        returns(true) implies (this@isAuthenticated is SymmetricKey.Authenticating<A, I, K>)
+        returns(false) implies (this@isAuthenticated is SymmetricKey.NonAuthenticating<I>)
+    }
+    return this.algorithm.authCapability is AuthCapability.Authenticated<*>
+}
+
+/**Use to smart cast*/
+@OptIn(ExperimentalContracts::class)
+fun <A : AuthCapability.Authenticated<*>, I : NonceTrait> SymmetricKey<A, I, *>.isIntegrated(): Boolean {
+    contract {
+        returns(true) implies (this@isIntegrated is SymmetricKey.Integrated.Authenticating<I>)
+        returns(false) implies (this@isIntegrated is SymmetricKey.WithDedicatedMac<I>)
+    }
+    return this.hasDedicatedMacKey()
+}
+
+/**Use to smart cast*/
+@OptIn(ExperimentalContracts::class)
+fun <A : AuthCapability<*>, I : NonceTrait> SymmetricKey<A, I, *>.hasDedicatedMacKey(): Boolean {
+    contract {
+        returns(true) implies (this@hasDedicatedMacKey is SymmetricKey.WithDedicatedMac<I>)
+        returns(false) implies (this@hasDedicatedMacKey is SymmetricKey.Integrated<A, I>)
+    }
+    return this is SymmetricKey.WithDedicatedMac
+}
+
+/**Use to smart cast*/
+@OptIn(ExperimentalContracts::class)
+fun <A : AuthCapability<K>, K : KeyType> SymmetricKey<A, *, K>.requiresNonce(): Boolean {
+    contract {
+        returns(true) implies (this@requiresNonce is SymmetricKey.RequiringNonce<A, K>)
+        returns(false) implies (this@requiresNonce is SymmetricKey.WithoutNonce<A, K>)
+    }
+    return algorithm.nonceTrait is NonceTrait.Required
+}
+
+/**
+ * The actual encryption key bytes
+ */
+val <A : AuthCapability<out KeyType.Integrated>, I : NonceTrait> SymmetricKey<A, I, out KeyType.Integrated>.secretKey get() = (this as SymmetricKey.Integrated).secretKey
+val <I : NonceTrait> SymmetricKey<AuthCapability.Authenticated.WithDedicatedMac<*,I>, I, out KeyType.WithDedicatedMacKey >.encryptionKey get() = (this as SymmetricKey.WithDedicatedMac).encryptionKey
```

Comment:
shouldn't at least one of these work simply on "keys"?

Reply:
i deliberately discrimante between secretKey and the duo of encryptionKey and macKey

Reply:
right, but this is a `ByteArray` property, isn't it? and "the encryption key" should always be possible to retrieve on any kind of cipher, shouldn't it?

Reply:
but it should not be called the same, if it requres a dedicated mac key, so you are made aware that tehre are two keys